hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java

```diff
@@ -50,21 +52,32 @@ private HiveSchemaConverter(boolean autoConvert) {
     this.id = 0;
   }
 
-  static Schema convert(List<String> names, List<TypeInfo> typeInfos, List<String> comments, boolean autoConvert) {
+  static Schema convert(List<String> names, List<TypeInfo> typeInfos, List<String> comments, boolean autoConvert,
+                        Set<String> identifierFieldSet) {
     HiveSchemaConverter converter = new HiveSchemaConverter(autoConvert);
-    return new Schema(converter.convertInternal(names, typeInfos, comments));
+    List<Types.NestedField> fields = converter.convertInternal(names, typeInfos, comments, identifierFieldSet);
+    Set<Integer> identifierIdSet = HiveSchemaUtil.identifierFieldIdSetOrThrow(fields, identifierFieldSet);
+    return new Schema(fields, identifierIdSet);
   }
 
   static Type convert(TypeInfo typeInfo, boolean autoConvert) {
     HiveSchemaConverter converter = new HiveSchemaConverter(autoConvert);
     return converter.convertType(typeInfo);
   }
 
-  List<Types.NestedField> convertInternal(List<String> names, List<TypeInfo> typeInfos, List<String> comments) {
+  List<Types.NestedField> convertInternal(List<String> names, List<TypeInfo> typeInfos, List<String> comments,
```

Comment:
Why not create a Schema here that has the identifier fields already converted, rather than recovering them by checking field names again?